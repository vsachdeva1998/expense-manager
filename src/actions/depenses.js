import database from '../firebase/firebase';

export const addDepense = (depense) => ({
  type: 'ADD_DEPENSE',
  depense,
});

export const startAddDepense = (depenseData = {}) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0,
    } = depenseData;
    const depense = { description, note, amount, createdAt };

    const ref = await database.ref(`users/${uid}/depenses`).push(depense);
    dispatch(
      addDepense({
        id: ref.key,
        ...depense,
      })
    );
  };
};

export const removeDepense = ({ id } = {}) => ({
  type: 'REMOVE_DEPENSE',
  id,
});

export const startRemoveDepense = ({ id } = {}) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;
    await database.ref(`users/${uid}/depenses/${id}`).remove();
    dispatch(removeDepense({ id }));
  };
};

export const editDepense = (id, updates) => ({
  type: 'EDIT_DEPENSE',
  id,
  updates,
});

export const startEditDepense = (id, updates) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;
    await database.ref(`users/${uid}/depenses/${id}`).update(updates);
    dispatch(editDepense(id, updates));
  };
};

export const setDepenses = (depenses) => ({
  type: 'SET_DEPENSES',
  depenses,
});

export const startSetDepenses = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;
    const snapshot = await database.ref(`users/${uid}/depenses`).once('value');
    const depenses = [];

    snapshot.forEach((childSnapshot) => {
      depenses.push({
        id: childSnapshot.key,
        ...childSnapshot.val(),
      });
    });
    dispatch(setDepenses(depenses));
  };
};
