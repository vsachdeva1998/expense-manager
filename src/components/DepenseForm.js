import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class DepenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.depense ? props.depense.description : '',
      note: props.depense ? props.depense.note : '',
      amount: props.depense ? (props.depense.amount / 100).toString() : '',
      createdAt: props.depense ? moment(props.depense.createdAt) : moment(),
      calendarFocused: false,
      error: '',
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and amount.',
      }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          type="text"
          className="text-input"
          placeholder="Description"
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          className="text-input"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          className="textarea"
          placeholder="Add a note for your depense"
          value={this.state.note}
          onChange={this.onNoteChange}
        ></textarea>
        <div>
          <button className="button">Save Depense</button>
        </div>
      </form>
    );
  }
}
