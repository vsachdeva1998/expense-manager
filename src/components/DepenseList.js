import React from 'react';
import { connect } from 'react-redux';
import DepenseListItem from './DepenseListItem';
import selectDepenses from '../selectors/depenses';

export const DepenseList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-small-screen">Depenses</div>
      <div className="show-big-screen">Depense</div>
      <div className="show-big-screen">Amount</div>
    </div>
    <div className="list-body">
      {props.depenses.length === 0 ? (
        <div className="list-item list-item--message">
          <span>No depenses</span>
        </div>
      ) : (
        props.depenses.map((depense) => (
          <DepenseListItem key={depense.id} {...depense} />
        ))
      )}
    </div>
  </div>
);

const mapStateToProps = (state) => ({
  depenses: selectDepenses(state.depenses, state.filters),
});

export default connect(mapStateToProps)(DepenseList);
