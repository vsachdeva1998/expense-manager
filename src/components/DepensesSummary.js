import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectDepenses from '../selectors/depenses';
import selectDepensesTotal from '../selectors/depenses-total';

export const DepensesSummary = ({ depenseCount, depensesTotal }) => {
  const depenseWord = depenseCount === 1 ? 'depense' : 'depenses';
  const formattedDepensesTotal = numeral(depensesTotal / 100).format('$0,0.00');
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing <span>{depenseCount}</span> {depenseWord} totalling{' '}
          <span>{formattedDepensesTotal}</span>
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">
            Add Depense
          </Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleDepenses = selectDepenses(state.depenses, state.filters);

  return {
    depenseCount: visibleDepenses.length,
    depensesTotal: selectDepensesTotal(visibleDepenses),
  };
};

export default connect(mapStateToProps)(DepensesSummary);
