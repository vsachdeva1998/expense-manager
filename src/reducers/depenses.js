const depensesReducerDefaultState = [];

export default (state = depensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_DEPENSE':
      return [...state, action.depense];
    case 'REMOVE_DEPENSE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_DEPENSE':
      return state.map((depense) => {
        if (depense.id === action.id) {
          return {
            ...depense,
            ...action.updates,
          };
        } else {
          return depense;
        }
      });
    case 'SET_DEPENSES':
      return action.depenses;
    default:
      return state;
  }
};
