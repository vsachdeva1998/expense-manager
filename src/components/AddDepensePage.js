import React from 'react';
import { connect } from 'react-redux';
import { startAddDepense } from '../actions/depenses';
import DepenseForm from './DepenseForm';

export class AddDepensePage extends React.Component {
  onSubmit = (depense) => {
    this.props.startAddDepense(depense);
    this.props.history.push('/dashboard');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Depense</h1>
          </div>
        </div>
        <div className="content-container">
          <DepenseForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddDepense: (depense) => dispatch(startAddDepense(depense)),
});

export default connect(undefined, mapDispatchToProps)(AddDepensePage);
