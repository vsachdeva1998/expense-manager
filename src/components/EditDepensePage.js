import React from 'react';
import { connect } from 'react-redux';
import { startEditDepense, startRemoveDepense } from '../actions/depenses';
import DepenseForm from './DepenseForm';

export class EditDepensePage extends React.Component {
  onSubmit = (depense) => {
    this.props.startEditDepense(this.props.depense.id, depense);
    this.props.history.push('/dashboard');
  };
  onRemove = () => {
    this.props.startRemoveDepense(this.props.depense.id);
    this.props.history.push('/dashboard');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Depense</h1>
          </div>
        </div>
        <div className="content-container">
          <DepenseForm depense={this.props.depense} onSubmit={this.onSubmit} />
          <button className="button button--secondary" onClick={this.onRemove}>
            Remove Depense
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  depense: state.depenses.find(
    (depense) => depense.id === props.match.params.id
  ),
});

const mapDispatchToProps = (dispatch) => ({
  startEditDepense: (id, depense) => dispatch(startEditDepense(id, depense)),
  startRemoveDepense: (id) => dispatch(startRemoveDepense({ id })),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditDepensePage);
