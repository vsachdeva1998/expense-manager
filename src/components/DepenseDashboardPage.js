import React from 'react';
import DepenseList from './DepenseList';
import DepenseListFilters from './DepenseListFilters';
import DepensesSummary from './DepensesSummary';

const DepenseDashboardPage = () => (
  <div>
    <DepensesSummary />
    <DepenseListFilters />
    <DepenseList />
  </div>
);

export default DepenseDashboardPage;
