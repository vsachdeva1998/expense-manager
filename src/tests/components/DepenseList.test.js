import React from 'react';
import { shallow } from 'enzyme';
import { DepenseList } from '../../components/DepenseList';
import depenses from '../fixtures/depenses';

test('should render DepenseList with depenses', () => {
  const wrapper = shallow(<DepenseList depenses={depenses} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render DepenseList with empty message', () => {
  const wrapper = shallow(<DepenseList depenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});
