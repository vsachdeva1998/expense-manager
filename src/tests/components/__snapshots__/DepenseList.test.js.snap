// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DepenseList with depenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-small-screen"
    >
      Depenses
    </div>
    <div
      className="show-big-screen"
    >
      Depense
    </div>
    <div
      className="show-big-screen"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <DepenseListItem
      amount={195}
      createdAt={0}
      description="Gum"
      id="1"
      key="1"
      note=""
    />
    <DepenseListItem
      amount={109500}
      createdAt={-345600000}
      description="Rent"
      id="2"
      key="2"
      note=""
    />
    <DepenseListItem
      amount={4500}
      createdAt={345600000}
      description="Credit Card"
      id="3"
      key="3"
      note=""
    />
  </div>
</div>
`;

exports[`should render DepenseList with empty message 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-small-screen"
    >
      Depenses
    </div>
    <div
      className="show-big-screen"
    >
      Depense
    </div>
    <div
      className="show-big-screen"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No depenses
      </span>
    </div>
  </div>
</div>
`;
